core/node/state_keeper/src/io/seal_logic/mod.rs

```diff
@@ -109,6 +110,96 @@ impl UpdatesManager {
             current_l1_batch_number = self.l1_batch.number
         );
 
+        let progress = L1_BATCH_METRICS.start(L1BatchSealStage::FilterWrittenSlots);
```

Comment:
Dumb question: Can this be implemented as a separate `StateKeeperOutputHandler`? IMO, if it would work, it'd be cleaner. Or would it be too complex to reason about non-atomicity w/ L1 batch header insertion? (AFAIU, batch sealing is synchronous now, and `StateKeeperOutputHandler`s are executed in `OutputHandler` one by one, so there shouldn't be an issue of not having an expected `l1_batches` table entry. OTOH, non-atomicity could mean that the metadata calculator would see an L1 batch w/o tree writes yet added.)

Reply:
In general, I like the idea
> OTOH, non-atomicity could mean that the metadata calculator would see an L1 batch w/o tree writes yet added

But I'm not sure how to overcome this, for old batches tree writes won't be present, so we can poll tree_writes in metadata with some reasonable timeout if the batch is the last sealed one, WDYT?

Reply:
I've opened a separate PR with new output handler, please check https://github.com/matter-labs/zksync-era/pull/2018. Part with `wait_for_tree_writes` looks a bit clunky to me but I don't know how to improve it